presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/GroupByWithConstants.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Capture;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.AggregationNode;
+import com.facebook.presto.spi.plan.Assignments;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.tree.Literal;
+import com.google.common.collect.ImmutableList;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeConstantGroupingKeys;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.castToExpression;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.isExpression;
+
+/**
+ * Transforms:
+ * <pre>
+ * - GroupBy key1, <const_expr>, key2
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project <const_expr>
+ *    - GroupBy key1, key2
+ * </pre>
+ */
+public class GroupByWithConstants
+        implements Rule<AggregationNode>
+{
+    private static final Capture<ProjectNode> CHILD = Capture.newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN =
+            aggregation()
+            .with(source().matching(project().capturedAs(CHILD)));
```

Comment:
@kaikalur is there a way to check if a variable reference is a constant without hard-coding the type of child node? that would miss cases when the constant is introduced further down in the tree, or is the result of a filter. Not sure if we support that type of property/constraint derivation?

Reply:
We are always rewriting the plans through all the optimizations so maybe just doing this at the end (just before exchanges) will do it. And I think this is the way to capture the children.